src/dstruct.d

```diff
@@ -238,22 +238,27 @@ extern (C++) class StructDeclaration : AggregateDeclaration
     // For those, today TypeInfo_Struct is generated in COMDAT.
     bool requestTypeInfo;
 
-    final extern (D) this(Loc loc, Identifier id)
+    final extern (D) this(Loc loc, Identifier id, bool inObject)
     {
         super(loc, id);
         zeroInit = 0; // assume false until we do semantic processing
         ispod = ISPODfwd;
+
         // For forward references
         type = new TypeStruct(this);
-        if (id == Id.ModuleInfo && !Module.moduleinfo)
-            Module.moduleinfo = this;
+
+        if (inObject)
+        {
+            if (id == Id.ModuleInfo && !Module.moduleinfo)
```

Comment:
When only considering my use-case, I don't think that's strictly needed.

I only need it to extend as far as making sure that I have the correct `ModuleInfo` and not some overridden type when it comes to laying out data.

I don't see much the point of reserving a list of names for user defined types myself, but I'd rather hand that batten to you.